packages/tangle/approvalweightmanager.go

```diff
@@ -0,0 +1,1431 @@
+		currentMarker := markers.NewMarker(marker.SequenceID(), i)
+		branchID := a.tangle.Booker.MarkersManager.BranchID(currentMarker)
+		if branchID != ledgerstate.MasterBranchID && a.Events.BranchConfirmation.Level(branchID) == 0 {
+			break
+		}
+
+		supportersOfMarker := a.supportersManager.SupportersOfMarker(currentMarker)
+		supporterWeight := float64(0)
+		if branchID == ledgerstate.MasterBranchID {
+			supportersOfMarker.ForEach(func(supporter Supporter) {
+				supporterWeight += activeWeights[supporter]
+			})
+		} else {
+			a.supportersManager.SupportersOfBranch(branchID).ForEach(func(supporter Supporter) {
+				if supportersOfMarker.Has(supporter) {
+					supporterWeight += activeWeights[supporter]
+				}
+			})
+		}
+
+		if _, transition := a.Events.MarkerConfirmation.Set(*currentMarker, supporterWeight/totalWeight); transition != events.ThresholdLevelIncreased {
+			break
+		}
+
+		a.lastConfirmedMarkers[marker.SequenceID()] = currentMarker.Index()
+	}
+}
+
+// lowestLastConfirmedMarker is an internal utility function that returns the last confirmed or lowest marker in a sequence.
+func (a *ApprovalWeightManager) lowestLastConfirmedMarker(sequenceID markers.SequenceID) (index markers.Index) {
+	index = a.lastConfirmedMarkers[sequenceID] + 1
+	a.tangle.Booker.MarkersManager.Manager.Sequence(sequenceID).Consume(func(sequence *markers.Sequence) {
+		if index < sequence.LowestIndex() {
+			index = sequence.LowestIndex()
+		}
+	})
+
+	return
+}
+
+func (a *ApprovalWeightManager) addMessageWeightToBranch(branchID ledgerstate.BranchID, message *Message) {
+	weightOfMessage, totalWeight := a.tangle.WeightProvider.Weight(
+		a.tangle.WeightProvider.Epoch(message.IssuingTime()),
+		message,
+	)
+
+	a.tangle.Storage.BranchWeight(branchID, func() *BranchWeight {
+		return NewBranchWeight(branchID)
+	}).Consume(func(branchWeight *BranchWeight) {
+		a.Events.BranchConfirmation.Set(branchID, branchWeight.IncreaseWeight(weightOfMessage/totalWeight)-a.weightOfHeaviestConflictingBranch(branchID))
+	})
+}
+
+func (a *ApprovalWeightManager) onBranchSupportRemoved(branchID ledgerstate.BranchID, message *Message) {
+	weightOfSupporter, totalWeight := a.tangle.WeightProvider.Weight(a.tangle.WeightProvider.Epoch(message.IssuingTime()), message)
+
+	a.tangle.Storage.BranchWeight(branchID, func() *BranchWeight {
+		return NewBranchWeight(branchID)
+	}).Consume(func(branchWeight *BranchWeight) {
+		a.Events.BranchConfirmation.Set(branchID, branchWeight.DecreaseWeight(weightOfSupporter/totalWeight)-a.weightOfHeaviestConflictingBranch(branchID))
+	})
+}
+
+func (a *ApprovalWeightManager) weightOfHeaviestConflictingBranch(branchID ledgerstate.BranchID) (weight float64) {
+	a.tangle.LedgerState.BranchDAG.ForEachConflictingBranchID(branchID, func(conflictingBranchID ledgerstate.BranchID) {
+		a.tangle.Storage.BranchWeight(conflictingBranchID).Consume(func(branchWeight *BranchWeight) {
+			weight = branchWeight.Weight()
```

Comment:
Why would this be the heaviest weight of conflicting branches? are these branch objects ordered by weight already?